File: py/classic/greedy_algorithm/ActivitySelection.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import functools
from typing import List

class ActivitySelection:
    def solution(self, values: List[List[int]]) -> int:
        values.sort(key = functools.cmp_to_key(self.__cmp))

        end, result = values[0][1], 1
        for value in values[1 : ]:
            if end <= value[0]:
                end = value[1]
                result += 1
        return result

    def __cmp(self, node1: List[int], node2: List[int]) -> int:
        return node1[1] - node2[1]

File: cpp/classic/greedy_algorithm/ActivitySelection.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class ActivitySelection {
public:
    int solution(vector<vector<int>> &values) {
        sort(values.begin(), values.end(), cmp);

        int end = values[0][1];
        int result = 1;
        for (int i = 1; i < (int) values.size(); ++i) {
            if (end <= values[i][0]) {
                end = values[i][1];
                ++result;
            }
        }
        return result;
    }

private:
    static bool cmp(const vector<int> &node1, const vector<int> &node2) {
        return node1[1] < node2[1];
    }
};
